File: backend/src/test/java/br/ufpr/sept/so2/modules/iam/application/LoginUseCaseTest.java
package br.ufpr.sept.so2.modules.iam.application;

import br.ufpr.sept.so2.modules.iam.domain.Usuario;
import br.ufpr.sept.so2.shared.domain.exception.CredenciaisInvalidasException;
import br.ufpr.sept.so2.shared.domain.valueobject.Email;
import br.ufpr.sept.so2.shared.domain.valueobject.Grr;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;

import java.time.OffsetDateTime;
import java.util.List;
import java.util.UUID;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertFalse;
import static org.junit.jupiter.api.Assertions.assertThrows;
import static org.junit.jupiter.api.Assertions.assertTrue;

class LoginUseCaseTest {

    private IamFakes.Usuarios usuarios;
    private IamFakes.Hasher hasher;
    private IamFakes.Audit audit;
    private LoginUseCase useCase;
    private Usuario aluno;

    @BeforeEach
    void setUp() {
        usuarios = new IamFakes.Usuarios();
        hasher = new IamFakes.Hasher();
        audit = new IamFakes.Audit();
        useCase = new LoginUseCase(
                usuarios,
                new IamFakes.RefreshTokens(),
                hasher,
                new IamFakes.Jwt(),
                new IamFakes.TokenHasher(),
                audit,
                new IamFakes.Settings()
        );
        aluno = usuario(true);
        usuarios.save(aluno);
    }

    @Test
    void loginComEmailOuGrrEmiteToken() {
        LoginResult porEmail = useCase.execute("aluno.dev@ufpr.br", "SenhaForte1!", "127.0.0.1");
        assertFalse(porEmail.mustChangePassword());
        assertTrue(porEmail.accessToken().startsWith("access:"));
        assertEquals(900, porEmail.expiresIn());

        LoginResult porGrr = useCase.execute("GRR20240001", "SenhaForte1!", "127.0.0.1");
        assertFalse(porGrr.mustChangePassword());
        assertTrue(audit.tipos.contains("iam.login_success"));
    }

    @Test
    void primeiroAcessoSinalizaMustChangePassword() {
        Usuario novo = usuario(false);
        usuarios.save(novo);
        LoginResult result = useCase.execute("novo.dev@ufpr.br", "SenhaForte1!", "10.0.0.1");
        assertTrue(result.mustChangePassword());
    }

    @Test
    void credenciaisInvalidasSaoIdenticasEUsamHashDummy() {
        assertThrows(CredenciaisInvalidasException.class,
                () -> useCase.execute("sumido@ufpr.br", "errada", "127.0.0.1"));
        assertThrows(CredenciaisInvalidasException.class,
                () -> useCase.execute("aluno.dev@ufpr.br", "errada", "127.0.0.1"));
        assertTrue(hasher.dummyCalls >= 1);
        assertTrue(audit.tipos.contains("iam.login_failed"));
    }

    @Test
    void decimaFalhaBloqueiaContaComEventoInterno() {
        for (int i = 0; i < 10; i++) {
            assertThrows(CredenciaisInvalidasException.class,
                    () -> useCase.execute("aluno.dev@ufpr.br", "errada", "127.0.0.1"));
        }
        assertTrue(audit.tipos.contains("iam.account_blocked"));
        assertTrue(usuarios.findById(aluno.getId()).orElseThrow().estaBloqueado(OffsetDateTime.now()));
    }

    private static Usuario usuario(boolean senhaAlterada) {
        OffsetDateTime agora = OffsetDateTime.parse("2026-01-01T00:00:00Z");
        String email = senhaAlterada ? "aluno.dev@ufpr.br" : "novo.dev@ufpr.br";
        String grr = senhaAlterada ? "GRR20240001" : "GRR20240002";
        return new Usuario(
                UUID.randomUUID(),
                Email.of(email),
                null,
                Grr.of(grr),
                "h:SenhaForte1!",
                senhaAlterada,
                senhaAlterada ? agora : null,
                null,
                null,
                true,
                0,
                null,
                List.of("dashboard.view_own"),
                agora,
                agora
        );
    }
}

File: backend/src/test/kotlin/br/ufpr/sept/so2/modules/iam/application/LoginUseCaseTest.kt
package br.ufpr.sept.so2.modules.iam.application

import br.ufpr.sept.so2.modules.iam.domain.Usuario
import br.ufpr.sept.so2.shared.domain.exception.CredenciaisInvalidasException
import br.ufpr.sept.so2.shared.domain.valueobject.Email
import br.ufpr.sept.so2.shared.domain.valueobject.Grr
import io.kotest.assertions.throwables.shouldThrow
import io.kotest.core.spec.style.StringSpec
import io.kotest.matchers.collections.shouldContain
import io.kotest.matchers.shouldBe
import io.kotest.matchers.string.shouldStartWith
import java.time.OffsetDateTime
import java.util.UUID

class LoginUseCaseTest : StringSpec({
    lateinit var usuarios: IamFakes.Usuarios
    lateinit var hasher: IamFakes.Hasher
    lateinit var audit: IamFakes.Audit
    lateinit var useCase: LoginUseCase
    lateinit var aluno: Usuario

    beforeEach {
        usuarios = IamFakes.Usuarios()
        hasher = IamFakes.Hasher()
        audit = IamFakes.Audit()
        useCase = LoginUseCase(
            usuarios,
            IamFakes.RefreshTokens(),
            hasher,
            IamFakes.Jwt(),
            IamFakes.TokenHasher(),
            audit,
            IamFakes.Settings(),
        )
        aluno = usuario(true)
        usuarios.save(aluno)
    }

    "login com e-mail ou GRR emite token" {
        val porEmail = useCase.execute("aluno.dev@ufpr.br", "SenhaForte1!", "127.0.0.1")
        porEmail.mustChangePassword shouldBe false
        porEmail.accessToken shouldStartWith "access:"
        porEmail.expiresIn shouldBe 900

        val porGrr = useCase.execute("GRR20240001", "SenhaForte1!", "127.0.0.1")
        porGrr.mustChangePassword shouldBe false
        audit.tipos shouldContain "iam.login_success"
    }

    "primeiro acesso sinaliza mustChangePassword" {
        val novo = usuario(false)
        usuarios.save(novo)
        val result = useCase.execute("novo.dev@ufpr.br", "SenhaForte1!", "10.0.0.1")
        result.mustChangePassword shouldBe true
    }

    "credenciais inválidas são idênticas e usam hash dummy" {
        shouldThrow<CredenciaisInvalidasException> {
            useCase.execute("sumido@ufpr.br", "errada", "127.0.0.1")
        }
        shouldThrow<CredenciaisInvalidasException> {
            useCase.execute("aluno.dev@ufpr.br", "errada", "127.0.0.1")
        }
        (hasher.dummyCalls >= 1) shouldBe true
        audit.tipos shouldContain "iam.login_failed"
    }

    "décima falha bloqueia conta com evento interno" {
        repeat(10) {
            shouldThrow<CredenciaisInvalidasException> {
                useCase.execute("aluno.dev@ufpr.br", "errada", "127.0.0.1")
            }
        }
        audit.tipos shouldContain "iam.account_blocked"
        usuarios.findById(aluno.id).orElseThrow().estaBloqueado(OffsetDateTime.now()) shouldBe true
    }
})

private fun usuario(senhaAlterada: Boolean): Usuario {
    val agora = OffsetDateTime.parse("2026-01-01T00:00:00Z")
    val email = if (senhaAlterada) "aluno.dev@ufpr.br" else "novo.dev@ufpr.br"
    val grr = if (senhaAlterada) "GRR20240001" else "GRR20240002"
    return Usuario(
        UUID.randomUUID(),
        Email.of(email),
        null,
        Grr.of(grr),
        "h:SenhaForte1!",
        senhaAlterada,
        if (senhaAlterada) agora else null,
        null,
        null,
        true,
        0,
        null,
        listOf("dashboard.view_own"),
        agora,
        agora,
    )
}
